core/src/main/java/io/grpc/internal/Http2ClientStreamTransportState.java

```diff
@@ -0,0 +1,236 @@
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+import java.nio.charset.Charset;
+
+import javax.annotation.Nullable;
+
+/**
+ * Base implementation for client streams using HTTP2 as the transport.
+ */
+public abstract class Http2ClientStreamTransportState extends AbstractClientStream2.TransportState {
+
+  /**
+   * Metadata marshaller for HTTP status lines.
+   */
+  private static final Metadata.AsciiMarshaller<Integer> HTTP_STATUS_LINE_MARSHALLER =
+      new Metadata.AsciiMarshaller<Integer>() {
+        @Override
+        public String toAsciiString(Integer value) {
+          return value.toString();
+        }
+
+        @Override
+        public Integer parseAsciiString(String serialized) {
+          return Integer.parseInt(serialized.split(" ", 2)[0]);
+        }
+      };
+
+  private static final Metadata.Key<Integer> HTTP2_STATUS = Metadata.Key.of(":status",
+      HTTP_STATUS_LINE_MARSHALLER);
+
+  /** When non-{@code null}, {@link #transportErrorMetadata} must also be non-{@code null}. */
+  private Status transportError;
+  private Metadata transportErrorMetadata;
+  private Charset errorCharset = Charsets.UTF_8;
+  private boolean contentTypeChecked;
+
+  protected Http2ClientStreamTransportState(int maxMessageSize) {
+    super(maxMessageSize);
+  }
+
+  protected abstract void http2ProcessingFailed(Status status, Metadata trailers);
+
+  /**
+   * Called by subclasses whenever {@code Headers} are received from the transport.
+   *
+   * @param headers the received headers
+   */
+  protected void transportHeadersReceived(Metadata headers) {
```

Comment:
While it _can_ be final, I didn't think it should be. These are very similar methods to `AbstractClientStream2`'s `inboundHeadersReceived()`, `inboundDataReceived()`, `inboundTrailersReceived()` and `AbstractServerStream`'s `inboundDataReceived()`, none of which are final. This is because some implementations may want to override the methods to make them visible to friend classes. Today that isn't done, but only because of data type differences; the methods are redeclared, but they overload instead of override because the headers and bytebuf types are slightly different.